Rewrite traces with complex tensors

hameerabbasi opened the pull request:
This PR enables rewriting of complex tensor traces.


cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @Lucaskabela

amjames committed:
[Do Not Merge] Experimenting with native complex -> subclass rewrites.

amjames committed:
add changes from pair session

hameerabbasi committed:
Get `_conj` and `_neg_view` (somewhat) working.

hameerabbasi committed:
Undo debugging changes.

hameerabbasi committed:
Fix typos and repr.

hameerabbasi commented on torch/_functorch/_aot_autograd/descriptors.py:
```diff
@@ -440,6 +440,23 @@ def expr(self) -> str:
+    def is_buffer(self) -> bool:
+        return self.original_input.is_buffer()
+
+    def is_tangen(self) -> bool:
```
Done in https://github.com/pytorch/pytorch/pull/169832/commits/b47a6bb6907de7b82a9fac0e1985823840a5ea78

hameerabbasi committed:
Fix some test failures.

hameerabbasi committed:
Fix failing tests using `torch.resolve_neg` and `aten.alias`.

hameerabbasi committed:
Fix newly-introduced lints.

hameerabbasi committed:
Fix more lints.

ezyang commented on torch/_subclasses/complex_tensor/_ops/common.py:
```diff
@@ -186,7 +186,7 @@ def split_complex_arg(
 def split_complex_tensor(complex_tensor: ComplexTensor) -> tuple[Tensor, Tensor]:
     """Split a ComplexTensor into its real and imaginary parts."""
-    return complex_tensor.re, complex_tensor.im
+    return torch.resolve_neg(complex_tensor.re), torch.resolve_neg(complex_tensor.im)
```
I don't remember if this is intended to be a view or not, but because you `resolve_neg` here, this will force clones to the output when the complex tensor is conjugated. This could be undesirable if you're using this in a context where a view was supposed to happen.

hameerabbasi commented on torch/_subclasses/complex_tensor/_core.py:
```diff
@@ -17,20 +17,45 @@
 
-    _re: Tensor
-    _im: Tensor
+    _data: Tensor
```
In order to preserve view semantics (most prominently for `torch.view_as_real`/`torch.view_as_complex` and `ComplexTensor.as_interleaved`), we store a tensor that would be the result of `torch.view_as_real` applied on a complex tensor.

hameerabbasi committed:
Add a more detailed comment for config var.

hameerabbasi committed:
Make sure `aten.view_as_real` takes into account `is_conj` and `is_neg`.

hameerabbasi committed:
Move ordering of `ComplexWrapper` closer to subclass handling.

hameerabbasi commented on torch/_functorch/_aot_autograd/graph_compile.py:
```diff
@@ -166,7 +167,11 @@ def _create_wrappers_for_dispatch(needs_autograd: bool) -> list[CompilerWrapper]
     """
-    return [AOTDedupeWrapper(), AOTSyntheticBaseWrapper(trace_joint=needs_autograd)]
+    return [
+        ComplexWrapper(),
```
Moved closer to subclass handling in https://github.com/pytorch/pytorch/pull/169832/commits/4554a8680ace41cb4d6ff14df4fd586c16b8f58f.

hameerabbasi committed:
Fix typo.

bdhirsh commented on torch/_functorch/config.py:
```diff
@@ -425,6 +425,13 @@ def remote_autograd_cache_default() -> Optional[bool]:
+# `torch._subclasses.complex_tensor.ComplexTensor` by decomposing everything into
+# real-valued operations, passing through the regular pipeline as necessary,
+# then converting back to a regular tensor.
+enable_complex_wrapper: bool = False
```
If this config is default-off, then shouldn't we turn it on in the `test_complex.py` tests?

I noticed that when i set this flag to true, this test starts failing:
```
python test/dynamo/test_complex.py -k test_simple
```

error:
```
  File "/data/users/hirsheybar/new2/pytorch/test/dynamo/test_complex.py", line 20, in test_simple
    self.assertEqual(fn_c(a, b, x), sample_function(a, b, x))
  File "/data/users/hirsheybar/new2/pytorch/torch/_dynamo/test_case.py", line 113, in assertEqual
    return super().assertEqual(x, y, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/hirsheybar/new2/pytorch/torch/testing/_internal/common_utils.py", line 4352, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: The values for attribute 'dtype' do not match: torch.complex64 != torch.float32.
```

hameerabbasi committed:
Set appropriate config during tests; and ensure they pass.

hameerabbasi committed:
Move subclass back to non-interleaved component format.

hameerabbasi committed:
Linting.

hameerabbasi committed:
Ban mutations on inputs.

hameerabbasi committed:
Ban output-aliasing-input.

ezyang commented on test/dynamo/test_complex.py:
```diff
@@ -0,0 +1,97 @@
+        a = torch.randn(2, 2, dtype=torch.complex64)
+        with self.assertRaises(
+            torch._dynamo.exc.BackendCompilerFailed,
+            msg=r"For wrapped complex arg \d+, mutating data is not supported.",
```
I hope this graph breaks when fullgraph=False!

hameerabbasi replied:
This fails to compile even without the RuntimeError changes for inductor and `fullgraph=False`:

```python
Traceback (most recent call last):
  File "/Users/habbasi/Quansight/torch-build/.pixi/envs/py313/lib/python3.13/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/Users/habbasi/Quansight/torch-build/.pixi/envs/py313/lib/python3.13/unittest/case.py", line 651, in run
    self._callTestMethod(testMethod)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/Users/habbasi/Quansight/torch-build/.pixi/envs/py313/lib/python3.13/unittest/case.py", line 606, in _callTestMethod
    if method() is not None:
       ~~~~~~^^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/testing/_internal/common_utils.py", line 3448, in wrapper
    method(*args, **kwargs)
    ~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/test/dynamo/test_complex.py", line 73, in test_aliasing_semantics
    self.assertEqual(f(a.clone()), fn_c2(a.clone()))
                                   ~~~~~^^^^^^^^^^^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_dynamo/eval_frame.py", line 1062, in compile_wrapper
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_dynamo/output_graph.py", line 2975, in _call_user_compiler
    raise BackendCompilerFailed(
        self.compiler_fn, e, inspect.currentframe()
    ).with_traceback(e.__traceback__) from None
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_dynamo/output_graph.py", line 2950, in _call_user_compiler
    compiled_fn = compiler_fn(gm, example_inputs)
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_dynamo/repro/after_dynamo.py", line 156, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/__init__.py", line 2482, in __call__
    return compile_fx(
        model_,
    ...<2 lines>...
        compile_region_name=self.name,
    )
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_inductor/compile_fx.py", line 2743, in compile_fx
    return _maybe_wrap_and_compile_fx_main(
        model_,
    ...<4 lines>...
        compile_region_name=compile_region_name,
    )
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_inductor/compile_fx.py", line 2824, in _maybe_wrap_and_compile_fx_main
    return _compile_fx_main(
        model_,
    ...<4 lines>...
        compile_region_name=compile_region_name,
    )
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_inductor/compile_fx.py", line 3022, in _compile_fx_main
    return dynamo_common.aot_autograd(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...<9 lines>...
        compile_region_name=compile_region_name,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    )(model_, example_inputs_)
    ~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_dynamo/backends/common.py", line 123, in __call__
    cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_functorch/aot_autograd.py", line 1233, in aot_module_simplified
    aot_graph_capture = aot_stage1_graph_capture(aot_state, functional_call)
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_functorch/_aot_autograd/graph_compile.py", line 249, in aot_stage1_graph_capture
    aot_dispatch_base_graph(
    ~~~~~~~~~~~~~~~~~~~~~~~^
        flat_fn,
        ^^^^^^^^
    ...<3 lines>...
        fw_metadata=aot_state.fw_metadata,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_functorch/_aot_autograd/graph_capture.py", line 379, in aot_dispatch_base_graph
    copy_count = assert_functional_graph(fw_module.graph)
  File "/Users/habbasi/Quansight/torch-build/git/pytorch/torch/_functorch/_aot_autograd/functional_utils.py", line 595, in assert_functional_graph
    raise AssertionError(error)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError: n=copy__1, n.args[0]=alias_1, placeholders={arg1_1, arg0_1}, graph=graph():
    %arg0_1 : [num_users=3] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=2] = placeholder[target=arg1_1]
    %is_pinned : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%arg0_1,), kwargs = {})
    %is_pinned_1 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%arg0_1,), kwargs = {})
    %is_pinned_2 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%arg1_1,), kwargs = {})
    %alias : [num_users=4] = call_function[target=torch.ops.aten.alias.default](args = (%arg0_1,), kwargs = {})
    %alias_1 : [num_users=3] = call_function[target=torch.ops.aten.alias.default](args = (%arg1_1,), kwargs = {})
    %is_pinned_3 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias,), kwargs = {})
    %is_pinned_4 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias,), kwargs = {})
    %is_pinned_5 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_1,), kwargs = {})
    %alias_2 : [num_users=2] = call_function[target=torch.ops.aten.alias.default](args = (%alias,), kwargs = {})
    %alias_3 : [num_users=2] = call_function[target=torch.ops.aten.alias.default](args = (%alias_1,), kwargs = {})
    %unsqueeze : [num_users=1] = call_function[target=torch.ops.aten.unsqueeze.default](args = (%alias_2, 2), kwargs = {})
    %unsqueeze_1 : [num_users=1] = call_function[target=torch.ops.aten.unsqueeze.default](args = (%alias_3, 2), kwargs = {})
    %cat : [num_users=1] = call_function[target=torch.ops.aten.cat.default](args = ([%unsqueeze, %unsqueeze_1], -1), kwargs = {})
    %full : [num_users=3] = call_function[target=torch.ops.aten.full.default](args = ([2, 2], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %full_1 : [num_users=2] = call_function[target=torch.ops.aten.full.default](args = ([2, 2], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %is_pinned_6 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%full,), kwargs = {})
    %is_pinned_7 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%full,), kwargs = {})
    %is_pinned_8 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%full_1,), kwargs = {})
    %alias_4 : [num_users=1] = call_function[target=torch.ops.aten.alias.default](args = (%full,), kwargs = {})
    %alias_5 : [num_users=1] = call_function[target=torch.ops.aten.alias.default](args = (%full_1,), kwargs = {})
    %copy : [num_users=1] = call_function[target=torch.ops.aten.copy.default](args = (%alias_2, %alias_4), kwargs = {})
    %alias_6 : [num_users=6] = call_function[target=torch.ops.aten.alias.default](args = (%copy,), kwargs = {})
    %copy_1 : [num_users=1] = call_function[target=torch.ops.aten.copy.default](args = (%alias_3, %alias_5), kwargs = {})
    %alias_7 : [num_users=4] = call_function[target=torch.ops.aten.alias.default](args = (%copy_1,), kwargs = {})
    %is_pinned_9 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_6,), kwargs = {})
    %is_pinned_10 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_6,), kwargs = {})
    %is_pinned_11 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_7,), kwargs = {})
    %alias_8 : [num_users=0] = call_function[target=torch.ops.aten.alias.default](args = (%alias_6,), kwargs = {})
    %alias_9 : [num_users=0] = call_function[target=torch.ops.aten.alias.default](args = (%alias_7,), kwargs = {})
    %is_pinned_12 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_6,), kwargs = {})
    %is_pinned_13 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_6,), kwargs = {})
    %is_pinned_14 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%alias_7,), kwargs = {})
    %alias_10 : [num_users=1] = call_function[target=torch.ops.aten.alias.default](args = (%alias_6,), kwargs = {})
    %alias_11 : [num_users=1] = call_function[target=torch.ops.aten.alias.default](args = (%alias_7,), kwargs = {})
    %copy_ : [num_users=3] = call_function[target=torch.ops.aten.copy_.default](args = (%alias, %alias_10), kwargs = {})
    %copy__1 : [num_users=2] = call_function[target=torch.ops.aten.copy_.default](args = (%alias_1, %alias_11), kwargs = {})
    %is_pinned_15 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%copy_,), kwargs = {})
    %is_pinned_16 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%copy_,), kwargs = {})
    %is_pinned_17 : [num_users=0] = call_function[target=torch.ops.aten.is_pinned.default](args = (%copy__1,), kwargs = {})
    %alias_12 : [num_users=0] = call_function[target=torch.ops.aten.alias.default](args = (%copy_,), kwargs = {})
    %alias_13 : [num_users=0] = call_function[target=torch.ops.aten.alias.default](args = (%copy__1,), kwargs = {})
    return (cat,)

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"


To execute this test, run the following from the base repo dir:
    python test/dynamo/test_complex.py ComplexTests.test_aliasing_semantics

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

The test below succeeds with `fullgraph=False`. I'm adding an xfail test with `fullgraph=False` for this one, and a passing one for the one below.

hameerabbasi replied:
See https://github.com/pytorch/pytorch/pull/169832/commits/bed033c39244c8cf3e68d1319e7d9e78d857cbbb.

hameerabbasi committed:
Add tests for graph-breaks.

ezyang commented on torch/_subclasses/complex_tensor/_core.py:
```diff
@@ -20,18 +20,32 @@ class ComplexTensor(Tensor):
 
+        if real.is_neg():
+            neg_flag = not neg_flag
+            conj_flag = not conj_flag
```
Hol' up, how come? If the real portion of the complex tensor is negated, that doesn't make it conjugated!

hameerabbasi replied:
See https://github.com/pytorch/pytorch/pull/169832/commits/0858ebd53dfe870d15a366c40974e5a19e5ed6e0

ezyang commented on torch/_subclasses/complex_tensor/_core.py:
```diff
@@ -128,24 +157,26 @@ def __tensor_unflatten__(
-    def forward(ctx: FunctionCtx, real: Tensor, imag: Tensor) -> ComplexTensor:  # type: ignore[bad-override]
+    def forward(  # type: ignore[bad-override]
+        ctx: FunctionCtx, real: Tensor, imag: Tensor
+    ) -> ComplexTensor:
```
please avoid reformatting like this, it just makes human review harder for no good reason

hameerabbasi replied:
This was done by lintrunner. I expect some rules changed.

ezyang commented on torch/_subclasses/complex_tensor/_core.py:
```diff
@@ -58,40 +72,55 @@ def __new__(cls, real: Tensor, imag: Tensor) -> Self:
-        res._re = real.clone().detach()
-        res._im = imag.clone().detach()
+        res._re = real.detach()
+        res._im = imag.detach()
```
I'm confused about the neg/conj contract for the inner `_re` and `_im` tensors. It seems plausible but the exact invariants aren't spelled out anywhere

hameerabbasi replied:
https://github.com/pytorch/pytorch/pull/169832/commits/0858ebd53dfe870d15a366c40974e5a19e5ed6e0

ezyang commented on torch/_subclasses/complex_tensor/_core.py:
```diff
@@ -58,40 +72,55 @@ def __new__(cls, real: Tensor, imag: Tensor) -> Self:
     @property
     def im(self) -> Tensor:
-        return self._im
+        negate = self.is_neg() != self.is_conj()
```
I just want to call out that there's not a good use case for negative views of complex tensors. Negative views arise because when you have a conjugate view and then project out the imaginary part, you need to talk about negative there.

hameerabbasi replied:
See https://github.com/pytorch/pytorch/pull/169832/commits/0858ebd53dfe870d15a366c40974e5a19e5ed6e0

ezyang commented on torch/_subclasses/complex_tensor/_ops/common.py:
```diff
@@ -228,14 +228,14 @@ def register_binary_nonlinear(
-        out_dt, (a_r, a_i, b_r, b_i) = promote_tensors(a_r, a_i, b_r, b_i)
+        out_dt, (a, b) = promote_tensors(a, b)
+        a_r, a_i = split_complex_arg(a)
+        b_r, b_i = split_complex_arg(b)
```
OOC, why did this get interchanged?

hameerabbasi replied:
Just a small clean-up.

ezyang commented on torch/_subclasses/complex_tensor/_ops/common.py:
```diff
@@ -300,6 +300,25 @@ def _as_interleaved(arg: ComplexTensor | Any) -> Tensor | Any:
+        if func.overloadpacket == torch.ops.aten.complex:
+            from .._core import Complex
+
+            return Complex.apply(*args)
```
It is surprising to me that the autograd function here does anything, as torch dispatch mode typically runs after autograd

hameerabbasi replied:
Running tests without this results in failures like the following in gradcheck:

```python
Exception: gradcheck expects at least one input tensor to require gradient, but none of the them have requires_grad=True.
```

<details><summary>Related diff</summary>

```diff
diff --git a/torch/_subclasses/complex_tensor/_core.py b/torch/_subclasses/complex_tensor/_core.py
index c83fd000da1..b6251b86d63 100644
--- a/torch/_subclasses/complex_tensor/_core.py
+++ b/torch/_subclasses/complex_tensor/_core.py
@@ -114,8 +114,8 @@ class ComplexTensor(Tensor):
             layout=layout,
             requires_grad=False,
         )
-        res._re = real.detach()
-        res._im = imag.detach()
+        res._re = real
+        res._im = imag
         # See "NOTE conj/neg (hameerabbasi)"
         torch._C._set_conj(res, conj_flag)
         torch._C._set_neg(res, neg_flag)
@@ -166,7 +166,7 @@ class ComplexTensor(Tensor):
     def from_interleaved(t: Tensor) -> ComplexTensor:
         re = torch.real(t)
         im = torch.imag(t) if t.dtype.is_complex else torch.zeros_like(t)
-        return Complex.apply(re, im)
+        return ComplexTensor(re, im)
 
     def as_interleaved(self) -> Tensor:
         return torch.complex(self.re, self.im)
```
</details>

ezyang commented on torch/_subclasses/complex_tensor/_ops/aten.py:
```diff
@@ -712,8 +721,21 @@ def logical_not_impl(self: ComplexTensor, *args: Any, **kwargs: Any) -> torch.Te
+            "the resulting tensor will NOT alias the original."
+        )
+    warnings.warn(
+        "`view_as_real` doesn't respect aliasing semantics under `torch.compile`.",
```
This warning is lazy. Just make the original and new storages immutable using the functional tensor facilities and error only when the user tries to observe the aliasing e.g., via mutation.

ezyang commented on torch/_subclasses/complex_tensor/_ops/aten.py:
```diff
@@ -712,8 +721,21 @@ def logical_not_impl(self: ComplexTensor, *args: Any, **kwargs: Any) -> torch.Te
+    )
+    out = torch.stack([self._re, self._im], dim=-1)
+    if self.is_neg():
+        out = torch._neg_view(out)
```
Since stack doesn't alias, using `_neg_view` here is needless complication

hameerabbasi replied:
This one is a correctness/behaviour matching issue. Consider:

```python
>>> a = torch.asarray(1+3j)
>>> a = torch._neg_view(torch.asarray(1+3j))
>>> a.is_neg()
True
>>> torch.view_as_real(a).is_neg()
True
```

This could lead to weird edge-case incorrectness in (admittedly pathological) code like this:

```python
def f(a):
    return 2*a if torch.view_as_real(a).is_neg() else torch.zeros_like(a)
```

hameerabbasi replied:
See https://github.com/pytorch/pytorch/pull/169832/commits/15b241c58b8b3eee9e453d7947ec5098e3fc0f7f

hameerabbasi committed:
Add comments wrt the neg/conj flags.

hameerabbasi committed:
Add test to demonstrate `view_as_real`/`is_neg` behaviour.

ezyang reviewed (changes requested):
tag me explicitly when you're ready

hameerabbasi committed:
Use functionalization for cases where input/output is mutated.

hameerabbasi committed:
Simplify some `conj`/`neg` flag handling.

hameerabbasi committed:
Lint.

amjames commented on torch/_functorch/_aot_autograd/graph_capture.py:
```diff
@@ -236,6 +237,17 @@ def _prepare_graph_capture_tracing(
             )
         )
 
+    complex_wrapper = ComplexWrapper()
```
I think we will want to move this under the `else` branch above, if we are relying on the wrapper being applied below functionalization to avoid incorrectness in aliasing/mutation. If functionalization is disabled we should skip the wrapper.

amjames replied:
We should also not silently ignore `config.enable_complex_wrapper` if the `config.disable_functionalization` state is incompatible with it, so it should either be an error that the two settings are in conflict or warning that `enable_complex_wrapper` is being ignored due to the incompatibility. 

I think a warning would be sufficient as skipping the wrapper simply falls back to the current behavior which is to fallback for most ops.

That logic could be implemented here, by hoisting [this](https://github.com/pytorch/pytorch/pull/169832/changes#diff-6ff2ea4c2706e1aa3306f23a6caa435f458bbc4d0cb2fdd32c46de5d24ee30d2R1458) check to this scope, or by testing `config.disable_functionalization` where that check is currently being performed.

amjames commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1405,6 +1408,101 @@ def post_compile(
+            else:
+                self.wrapped_arg_indices.append(i)
+                complex_arg = ComplexTensor.from_interleaved(arg)
+                new_args.append(complex_arg.re)
```
Why wrap, then append `re/im` as separate slots? 

We are applying the wrapper above subclass desugaring, so that should unpack the wrapper subclass for us, after we have retraced with the subclass in place to dispatch to the un-interleaved op implementations.

The way this is implemented when we hit `aot_dispatch_subclass` [here](https://github.com/pytorch/pytorch/pull/169832/changes#diff-e55fc91932ff4cc67bb27fc47de413a96245364c6959c80c0d3e40fa363f7ea6L239) there will be no subclass in place

amjames commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1405,6 +1408,101 @@ def post_compile(
+
+        wrapped_args = pytree.tree_map(self.wrap, flat_args)
+        if len(self.wrapped_arg_indices) == 0:
+            return flat_fn, flat_args, flat_args_descs, fw_metadata
```
If we have no wrapped inputs we may still have wrapped intermediates on the graph and would want to provide the `wrapped_flat_fn` rather than `flat_fn` directly? 

eg with this function:
a, b are 2d float32.
```python
def f(a, b):
   c = torch.sin(a)
   d = torch.cos(b)
   x = torch.complex(a, b)
   y = torch.complex(c, d)
   z = torch.mm(x, y)
   return z.real + z.imag
```

This condition would mean that we do no wrapping of x, y and therefore the `mm`  will hit a fallback.

amjames commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1405,6 +1408,101 @@ def post_compile(
+            static_input_indices=aot_config.static_input_indices,
+            keep_input_mutations=fw_metadata.keep_input_mutations,
+        )(*wrapped_args)
+        aot_config._did_wrap_complex = True
```
nit: I think it would be more appropriate to put this on the updated metadata rather than the config object.

hameerabbasi committed:
Move around the complex wrapper triggers, do codegen for post_compile.

hameerabbasi commented on test/dynamo/test_complex.py:
```diff
@@ -0,0 +1,87 @@
+        a = torch.randn(2, 2, dtype=torch.complex64)
+        fn_c = torch.compile(f, fullgraph=True)
+        self.assertEqual(f(a.clone()), fn_c(a.clone()))
+
```
This test fails

hameerabbasi reviewed (commented):
@ezyang As discussed off-band, one test fails due to `torch.view_as_real` not being able to preserve view semantics. If that's acceptable, let's xfail the test, open an issue, and review the rest of the PR.

ezyang committed:
AI SLOP

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
file

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

hameerabbasi committed:
Simplify Ed's code.

ezyang reviewed (changes requested):
my current understanding is action is not currently needed from me, getting this out of my queue per https://github.com/pytorch/pytorch/wiki/Pull-request-review-conventions

hameerabbasi committed:
Fix tracing when none of the inputs/outputs are complex, but intermediates are.

hameerabbasi committed:
Check for incorrect aliasing.

hameerabbasi committed:
Add test for RoPE.

hameerabbasi committed:
Remove unnecessary code.

hameerabbasi committed:
Update warning for complex compilation.

ezyang committed:
Simplify ComplexTensor: remove conj/neg bits, aliasing checks, and experimental warning.

Code review of the complex decomposition pass identified several issues:

1. Metadata preservation during retrace: decompose_complex_in_graph retraces
   the graph via make_fx + Interpreter, but was not wrapping the call in
   fx_traceback.preserve_node_meta(). This meant stack_trace,
   nn_module_stack, source_fn_stack, and all other _COPY_META_FIELDS were
   silently dropped. val/tensor_meta were fine (freshly computed), but
   debuggability suffered. Fixed by wrapping make_fx in
   preserve_node_meta().

2. conj/neg bit tracking on ComplexTensor was broken dead code. The
   constructor lifted neg/conj flags from input real/imag tensors onto the
   wrapper, but never stripped them from the stored parts (detach doesn't
   clear them), and the re/im properties never consulted the wrapper's
   flags. In practice no code path fed tensors with neg/conj bits into the
   constructor, so this was harmless but misleading. Removed entirely --
   conj and neg are now decomposed at the op level into arithmetic on the
   real/imag parts (_conj -> negate imag, _neg_view -> negate both).

3. INCORRECT_ALIASING_OPS and _assert_no_incorrect_aliasing_mutation were
   unnecessary. The check guarded against view_as_real/view_as_complex
   producing copies (not views) under ComplexTensor, which could diverge
   from eager when the aliased storage was also mutated. But this pass runs
   post-functionalization -- there are no mutations in the graph -- so
   broken aliasing is harmless. Inductor re-establishes any required output
   aliasing. Removed the check, the INCORRECT_ALIASING_OPS set, and the
   register_incorrect_aliasing decorator.

4. The _warn_experimental_complex_wrapper warning was removed. When the
   complex wrapper is enabled, complex ops are decomposed into real ops
   before Inductor sees them, so the "performance may be worse" caveat
   does not apply.

5. Added test_input_mutation to exercise complex input mutation through
   the post-functionalization pipeline, which had zero test coverage.
   Changed test_incorrect_aliasing_semantics_raises to
   test_view_as_real_with_mutation, which now expects success.

Authored by Claude.

hameerabbasi committed:
Revert "Simplify ComplexTensor: remove conj/neg bits, aliasing checks, and experimental warning."

This reverts commit 11852e3252669a826b1a86705e41ad65f6e24d2e.

hameerabbasi committed:
Add back useful pieces after the reversion of 11852e3252669a826b1a86705e41ad65f6e24d2e.

1. Improve debuggability via `fx_traceback.preserve_node_meta()`.
2. Undo changes 2/3 from original commit message, as this could introduce silent incorrectness:
   * 2 removed `conj`/`neg` bit tracking which broke aliasing semantics for
     `aten.conj` and `aten._neg_view`. This was added back, with tests showing
     the desired behaviour in test/test_dynamo/test_complex.py::{test_conj_alias, test_neg_alias}.
   * 3 removed a check for incorrect aliasing ops; inputs/outputs of these
     needed to be checked for mutations. See `test_complex_output_aliases_input_raises`,
     `test_complex_input_mutation_raises`, and `test_incorrect_aliasing_semantics_raises`.
     Previously, these 3 tests would produce silent incorrectness.
3. Add back the warning removed by 4, but modify the wording to say missing ops rather
   than flagging performance concerns.
4. Refine the check for incorrectness; so that less `.clone()`s are needed. Remove some
   from the RoPE test.

hameerabbasi committed:
Remove handling of `conj`/`neg` flags.

hameerabbasi committed:
Input-alias-output works, and remove mentions of deleted note.

hameerabbasi committed:
Get rid of warning.

ezyang commented on torch/_functorch/_aot_autograd/complex_decomposition.py:
```diff
@@ -0,0 +1,171 @@
+    return seen
+
+
+def _assert_no_incorrect_aliasing_mutation(gm: fx.GraphModule) -> None:
```
There's probably a more integrated way to do this but whatever, self contained is fine too

ezyang commented on torch/_functorch/_aot_autograd/complex_decomposition.py:
```diff
@@ -0,0 +1,171 @@
+    for n in gm.graph.nodes:
+        if isinstance(n.target, torch._ops.OpOverload) and (
+            n.target.overloadpacket in INCORRECT_ALIASING_OPS
+            or n.target in INCORRECT_ALIASING_OPS
```
Let's simplify this code by universally banning ALL mutation on complex inputs

hameerabbasi replied:
Done in https://github.com/pytorch/pytorch/pull/169832/commits/2b7718faad291eb08d6340caf6044be81ae74761.

hameerabbasi committed:
Ban all mutations.
